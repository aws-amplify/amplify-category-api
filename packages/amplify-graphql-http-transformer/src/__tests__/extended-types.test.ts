import { testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { AppSyncAuthConfiguration } from '@aws-amplify/graphql-transformer-interfaces';
import { parse } from 'graphql';
import { HttpTransformer } from '..';

describe('@http directive on extended types', () => {
  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [{ authenticationType: 'AWS_IAM' }],
  };

  test.each(['Query', 'Mutation', 'Subscription'])('supports @http directive on fields of %s type extensions', (builtInType: string) => {
    const schema = /* GraphQL */ `
      type ${builtInType} {
        customOperation1: String! @http(url: "https://www.api.com/foo1")
      }

      extend type ${builtInType} {
        customOperation2: String! @http(url: "https://www.api.com/foo2")
      }
    `;

    const testTransformParams = {
      schema: schema,
      authConfig,
      transformers: [new HttpTransformer()],
    };

    const out = testTransform(testTransformParams);
    expect(out).toBeDefined();
    expect(out.resolvers).toMatchSnapshot();
    expect(out.pipelineFunctions).toMatchSnapshot();
    parse(out.schema);
  });

  test.each(['Query', 'Mutation'])('does not support @http directive on %s object extension itself', (builtInType: string) => {
    const schema = /* GraphQL */ `
        type ${builtInType} {
          customOperation1: String! @http(url: "https://www.api.com/foo1")
        }

        extend type ${builtInType} @http(url: "https://www.api.com/foo2") {
          customOperation2: String! 
        }
      `;

    const testTransformParams = {
      schema: schema,
      authConfig,
      transformers: [new HttpTransformer()],
    };

    // The GraphQL parser actually catches this case, but we'll leave it here for completeness
    expect(() => testTransform(testTransformParams)).toThrow('Directive "@http" may not be used on OBJECT.');
  });

  test('does not support @http directive on fields of non-model type extensions', () => {
    const schema = /* GraphQL */ `
      type Foo {
        customField1: String! @http(url: "https://www.api.com/foo1")
      }

      extend type Foo {
        customField2: String! @http(url: "https://www.api.com/foo2")
      }
    `;

    const testTransformParams = {
      schema: schema,
      authConfig,
      transformers: [new HttpTransformer()],
    };

    expect(() => testTransform(testTransformParams)).toThrow(
      "The '@http' directive cannot be used on fields of type extensions other than 'Query', 'Mutation', and 'Subscription'. See Foo.customField2",
    );
  });
});
