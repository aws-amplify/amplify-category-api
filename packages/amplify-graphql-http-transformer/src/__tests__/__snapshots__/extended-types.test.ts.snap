// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@http directive on extended types supports @http directive on fields of Mutation type extensions 1`] = `
Object {
  "Mutation.customOperation1.DataResolver.req.vtl": "## [Start] Create request. **
#set( $headers = $utils.http.copyHeaders($ctx.request.headers) )
$util.qr($headers.put(\\"accept-encoding\\", \\"application/json\\"))
{
  \\"version\\": \\"2018-05-29\\",
  \\"method\\": \\"GET\\",
  \\"resourcePath\\": \\"/foo1\\",
  \\"params\\": {
      \\"headers\\": $util.toJson($headers),
      \\"query\\": $util.toJson($ctx.args.query)
  }
}
## [End] Create request. **",
  "Mutation.customOperation1.DataResolver.res.vtl": "## [Start] Process response. **
#if( $ctx.result.statusCode == 200 )
  #if( $ctx.result.headers.get(\\"Content-Type\\").toLowerCase().contains(\\"xml\\") )
$utils.xml.toJsonString($ctx.result.body)
  #else
$ctx.result.body
  #end
#else
$util.qr($util.appendError($ctx.result.body, $ctx.result.statusCode))
#end
## [End] Process response. **",
  "Mutation.customOperation1.res.vtl": "$util.toJson($ctx.prev.result)",
  "Mutation.customOperation2.DataResolver.req.vtl": "## [Start] Create request. **
#set( $headers = $utils.http.copyHeaders($ctx.request.headers) )
$util.qr($headers.put(\\"accept-encoding\\", \\"application/json\\"))
{
  \\"version\\": \\"2018-05-29\\",
  \\"method\\": \\"GET\\",
  \\"resourcePath\\": \\"/foo2\\",
  \\"params\\": {
      \\"headers\\": $util.toJson($headers),
      \\"query\\": $util.toJson($ctx.args.query)
  }
}
## [End] Create request. **",
  "Mutation.customOperation2.DataResolver.res.vtl": "## [Start] Process response. **
#if( $ctx.result.statusCode == 200 )
  #if( $ctx.result.headers.get(\\"Content-Type\\").toLowerCase().contains(\\"xml\\") )
$utils.xml.toJsonString($ctx.result.body)
  #else
$ctx.result.body
  #end
#else
$util.qr($util.appendError($ctx.result.body, $ctx.result.statusCode))
#end
## [End] Process response. **",
  "Mutation.customOperation2.res.vtl": "$util.toJson($ctx.prev.result)",
}
`;

exports[`@http directive on extended types supports @http directive on fields of Mutation type extensions 2`] = `Object {}`;

exports[`@http directive on extended types supports @http directive on fields of Query type extensions 1`] = `
Object {
  "Query.customOperation1.DataResolver.req.vtl": "## [Start] Create request. **
#set( $headers = $utils.http.copyHeaders($ctx.request.headers) )
$util.qr($headers.put(\\"accept-encoding\\", \\"application/json\\"))
{
  \\"version\\": \\"2018-05-29\\",
  \\"method\\": \\"GET\\",
  \\"resourcePath\\": \\"/foo1\\",
  \\"params\\": {
      \\"headers\\": $util.toJson($headers),
      \\"query\\": $util.toJson($ctx.args.query)
  }
}
## [End] Create request. **",
  "Query.customOperation1.DataResolver.res.vtl": "## [Start] Process response. **
#if( $ctx.result.statusCode == 200 )
  #if( $ctx.result.headers.get(\\"Content-Type\\").toLowerCase().contains(\\"xml\\") )
$utils.xml.toJsonString($ctx.result.body)
  #else
$ctx.result.body
  #end
#else
$util.qr($util.appendError($ctx.result.body, $ctx.result.statusCode))
#end
## [End] Process response. **",
  "Query.customOperation1.res.vtl": "$util.toJson($ctx.prev.result)",
  "Query.customOperation2.DataResolver.req.vtl": "## [Start] Create request. **
#set( $headers = $utils.http.copyHeaders($ctx.request.headers) )
$util.qr($headers.put(\\"accept-encoding\\", \\"application/json\\"))
{
  \\"version\\": \\"2018-05-29\\",
  \\"method\\": \\"GET\\",
  \\"resourcePath\\": \\"/foo2\\",
  \\"params\\": {
      \\"headers\\": $util.toJson($headers),
      \\"query\\": $util.toJson($ctx.args.query)
  }
}
## [End] Create request. **",
  "Query.customOperation2.DataResolver.res.vtl": "## [Start] Process response. **
#if( $ctx.result.statusCode == 200 )
  #if( $ctx.result.headers.get(\\"Content-Type\\").toLowerCase().contains(\\"xml\\") )
$utils.xml.toJsonString($ctx.result.body)
  #else
$ctx.result.body
  #end
#else
$util.qr($util.appendError($ctx.result.body, $ctx.result.statusCode))
#end
## [End] Process response. **",
  "Query.customOperation2.res.vtl": "$util.toJson($ctx.prev.result)",
}
`;

exports[`@http directive on extended types supports @http directive on fields of Query type extensions 2`] = `Object {}`;

exports[`@http directive on extended types supports @http directive on fields of Subscription type extensions 1`] = `
Object {
  "Subscription.customOperation1.DataResolver.req.vtl": "## [Start] Create request. **
#set( $headers = $utils.http.copyHeaders($ctx.request.headers) )
$util.qr($headers.put(\\"accept-encoding\\", \\"application/json\\"))
{
  \\"version\\": \\"2018-05-29\\",
  \\"method\\": \\"GET\\",
  \\"resourcePath\\": \\"/foo1\\",
  \\"params\\": {
      \\"headers\\": $util.toJson($headers),
      \\"query\\": $util.toJson($ctx.args.query)
  }
}
## [End] Create request. **",
  "Subscription.customOperation1.DataResolver.res.vtl": "## [Start] Process response. **
#if( $ctx.result.statusCode == 200 )
  #if( $ctx.result.headers.get(\\"Content-Type\\").toLowerCase().contains(\\"xml\\") )
$utils.xml.toJsonString($ctx.result.body)
  #else
$ctx.result.body
  #end
#else
$util.qr($util.appendError($ctx.result.body, $ctx.result.statusCode))
#end
## [End] Process response. **",
  "Subscription.customOperation1.res.vtl": "$util.toJson($ctx.prev.result)",
  "Subscription.customOperation2.DataResolver.req.vtl": "## [Start] Create request. **
#set( $headers = $utils.http.copyHeaders($ctx.request.headers) )
$util.qr($headers.put(\\"accept-encoding\\", \\"application/json\\"))
{
  \\"version\\": \\"2018-05-29\\",
  \\"method\\": \\"GET\\",
  \\"resourcePath\\": \\"/foo2\\",
  \\"params\\": {
      \\"headers\\": $util.toJson($headers),
      \\"query\\": $util.toJson($ctx.args.query)
  }
}
## [End] Create request. **",
  "Subscription.customOperation2.DataResolver.res.vtl": "## [Start] Process response. **
#if( $ctx.result.statusCode == 200 )
  #if( $ctx.result.headers.get(\\"Content-Type\\").toLowerCase().contains(\\"xml\\") )
$utils.xml.toJsonString($ctx.result.body)
  #else
$ctx.result.body
  #end
#else
$util.qr($util.appendError($ctx.result.body, $ctx.result.statusCode))
#end
## [End] Process response. **",
  "Subscription.customOperation2.res.vtl": "$util.toJson($ctx.prev.result)",
}
`;

exports[`@http directive on extended types supports @http directive on fields of Subscription type extensions 2`] = `Object {}`;
